lib/spack/spack/traverse.py

```diff
@@ -96,18 +96,75 @@ def __init__(self, visitor, key=id, visited=None):
         self.visited = set() if visited is None else visited
         self.key = key
 
-    def accept(self, node):
-        return self.visitor.accept(node)
+    def accept(self, item):
+        return self.visitor.accept(item)
 
-    def neighbors(self, node):
+    def neighbors(self, item):
         # Covering edges means: drop dependencies of visited nodes.
-        key = self.key(node.edge.spec)
+        key = self.key(item.edge.spec)
 
         if key in self.visited:
             return []
 
         self.visited.add(key)
-        return self.visitor.neighbors(node)
+        return self.visitor.neighbors(item)
+
+
+class TopoVisitor(object):
+    """Visitor that can be used in :py:func:`depth-first traversal
+    <spack.traverse.traverse_depth_first_with_visitor>` to generate
+    a topologically ordered list of edges (in-edges before out-edges).
```

Comment:
Note that by itself, the constraint "in-edges before out-edges" does not produce a topological order:

```
0
|\
1 |
|/
3
```

`[0, 3], [0, 1], [1, 3]` satisfies that but is not a valid topological ordering. The comment should probably first introduce the traditional concept of topological ordering and then extend it to edges.

This is partly my concern with generalizing the notion of topological ordering to edges: suddenly the documentation speaks in terms that are unfamiliar and need to be reasoned about, vs. concepts which are immediately accessible (e.g. via a Wikipedia article) - reorganizing the docstring would help with this though.

Reply:
I agree. I'm not sure how useful it is to allow cover=edges at this point. You do have the property that `map(lambda s: s.parent, traverse_edges(order="topo))` is topo ordered, but yeah, that does not include all reachable vertices.

Reply:
Sorry to be clear I trusted that this produces a topological order (or that one can be extracted from the list of edges), my concern is that it will be hard for future maintainers to understand it; the invariant stated in the docstring is insufficient to guarantee a topological ordering (the algorithm itself does more than that though), which I think is an example of how it's challenging to describe the semantics.

Are you inclined to undo the support for `cover=edges`, or would you prefer to improve the docstrings? Either way is fine with me.

Reply:
I just updated the docstrings and kept `cover=edges`